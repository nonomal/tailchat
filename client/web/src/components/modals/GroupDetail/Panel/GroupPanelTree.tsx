import React, { useCallback, useMemo } from 'react';
import {
  deleteGroupPanel,
  GroupPanel as GroupPanelInfo,
  showAlert,
  t,
} from 'tailchat-shared';
import { Space, Tree } from 'antd';
import type { DataNode } from 'antd/lib/tree';
import { buildTreeDataWithGroupPanel } from './utils';
import { useGroupPanelTreeDrag } from './useGroupPanelTreeDrag';
import { closeModal, openModal } from '@/components/Modal';
import { ModalModifyGroupPanel } from '../../GroupPanel/ModifyGroupPanel';
import { IconBtn } from '@/components/IconBtn';

interface GroupPanelTree {
  groupId: string;
  groupPanels: GroupPanelInfo[];
  onChange: (newGroupPanels: GroupPanelInfo[]) => void;
}
/**
 * 仅同层级相互拉的树形结构
 */
export const GroupPanelTree: React.FC<GroupPanelTree> = React.memo((props) => {
  const treeData: DataNode[] = useMemo(
    () => buildTreeDataWithGroupPanel(props.groupPanels),
    [props.groupPanels]
  );

  const handleModifyPanel = useCallback(
    (panelId: string) => {
      const key = openModal(
        <ModalModifyGroupPanel
          groupId={props.groupId}
          groupPanelId={panelId}
          onSuccess={() => closeModal(key)}
        />
      );
    },
    [props.groupId]
  );

  const handleDeletePanel = useCallback(
    (panelId: string, panelName: string, isGroup: boolean) => {
      showAlert({
        message: isGroup
          ? t('确定要删除面板组 【{{name}}】 以及下级的所有面板么', {
              name: panelName,
            })
          : t('确定要删除面板 【{{name}}】 么', { name: panelName }),
        onConfirm: async () => {
          await deleteGroupPanel(props.groupId, panelId);
        },
      });
    },
    [props.groupId]
  );

  const titleRender = useCallback(
    (node: DataNode): React.ReactNode => {
      return (
        <div className="flex group">
          <span>{node.title as string}</span>
          <div className="opacity-0 group-hover:opacity-100 ml-2">
            <Space size="small">
              <IconBtn
                title={t('编辑')}
                type="text"
                size="small"
                icon="mdi:pencil-outline"
                onClick={(e) => {
                  e.stopPropagation();
                  e.preventDefault();
                  handleModifyPanel(String(node.key));
                }}
              />

              <IconBtn
                title={t('删除')}
                type="text"
                size="small"
                icon="mdi:trash-can-outline"
                onClick={(e) => {
                  e.stopPropagation();
                  e.preventDefault();
                  handleDeletePanel(
                    String(node.key),
                    String(node.title),
                    !node.isLeaf
                  );
                }}
              />
            </Space>
          </div>
        </div>
      );
    },
    [handleDeletePanel]
  );

  const { handleDragStart, handleDragEnd, handleAllowDrop, handleDrop } =
    useGroupPanelTreeDrag(props.groupPanels, props.onChange);

  return (
    <Tree
      treeData={treeData}
      defaultExpandAll={true}
      blockNode={true}
      draggable={{
        icon: false,
      }}
      selectable={false}
      titleRender={titleRender}
      onDrop={handleDrop}
      // TODO: 待简化 https://github.com/react-component/tree/pull/482
      onDragStart={handleDragStart}
      onDragEnd={handleDragEnd}
      allowDrop={handleAllowDrop}
    />
  );
});
GroupPanelTree.displayName = 'GroupPanelTree';
