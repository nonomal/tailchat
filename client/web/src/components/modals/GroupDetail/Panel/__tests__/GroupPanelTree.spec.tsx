import { render } from '@testing-library/react';
import React from 'react';
import { GroupPanel, GroupPanelType } from 'tailchat-shared';
import { GroupPanelTree } from '../GroupPanelTree';

const testGroupPanels: GroupPanel[] = [
  {
    id: '00',
    name: 'section-1',
    type: GroupPanelType.GROUP,
  },
  {
    id: '01',
    name: 'panel-01',
    type: GroupPanelType.TEXT,
    parentId: '00',
  },
  {
    id: '02',
    name: 'panel-02',
    type: GroupPanelType.TEXT,
    parentId: '00',
  },
  {
    id: '10',
    name: 'section-2',
    type: GroupPanelType.GROUP,
  },
  {
    id: '11',
    name: 'panel-11',
    type: GroupPanelType.TEXT,
    parentId: '10',
  },
  {
    id: '12',
    name: 'panel-12',
    type: GroupPanelType.TEXT,
    parentId: '10',
  },
];

describe('GroupPanelTree', () => {
  test('simple render snapshot', async () => {
    const onChange = jest.fn();
    const wrapper = render(
      <GroupPanelTree
        groupId={'fakeGroupId'}
        groupPanels={testGroupPanels}
        onChange={onChange}
      />
    );

    expect(wrapper.container).toMatchSnapshot();
  });
});
