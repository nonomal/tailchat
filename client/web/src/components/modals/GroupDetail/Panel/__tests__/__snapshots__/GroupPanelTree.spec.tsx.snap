// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GroupPanelTree simple render snapshot 1`] = `
<div>
  <div
    class="ant-tree ant-tree-icon-hide ant-tree-block-node ant-tree-unselectable"
    role="tree"
  >
    <div>
      <input
        aria-label="for screen reader"
        style="width: 0px; height: 0px; display: flex; overflow: hidden; opacity: 0; border: 0px; padding: 0px; margin: 0px;"
        tabindex="0"
        value=""
      />
    </div>
    <div
      aria-hidden="true"
      class="ant-tree-treenode"
      style="position: absolute; pointer-events: none; visibility: hidden; height: 0px; overflow: hidden; border: 0px; padding: 0px;"
    >
      <div
        class="ant-tree-indent"
      >
        <div
          class="ant-tree-indent-unit"
        />
      </div>
    </div>
    <div
      class="ant-tree-list"
      style="position: relative;"
    >
      <div
        class="ant-tree-list-holder"
      >
        <div>
          <div
            class="ant-tree-list-holder-inner"
            style="display: flex; flex-direction: column;"
          >
            <div
              aria-grabbed="false"
              class="ant-tree-treenode ant-tree-treenode-switcher-open ant-tree-treenode-draggable"
              draggable="true"
            >
              <span
                aria-hidden="true"
                class="ant-tree-indent"
              />
              <span
                class="ant-tree-switcher ant-tree-switcher_open"
              >
                <span
                  aria-label="caret-down"
                  class="anticon anticon-caret-down ant-tree-switcher-icon"
                  role="img"
                >
                  <svg
                    aria-hidden="true"
                    data-icon="caret-down"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="0 0 1024 1024"
                    width="1em"
                  >
                    <path
                      d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                    />
                  </svg>
                </span>
              </span>
              <span
                class="ant-tree-node-content-wrapper ant-tree-node-content-wrapper-open"
                title="section-1"
              >
                <span
                  class="ant-tree-title"
                >
                  <div
                    class="flex group"
                  >
                    <span>
                      section-1
                    </span>
                    <div
                      class="opacity-0 group-hover:opacity-100 ml-2"
                    >
                      <div
                        class="ant-space ant-space-horizontal ant-space-align-center"
                      >
                        <div
                          class="ant-space-item"
                          style="margin-right: 8px;"
                        >
                          <button
                            class="ant-btn ant-btn-circle ant-btn-text ant-btn-sm ant-btn-icon-only border-0 text-white text-opacity-80 hover:text-opacity-100 bg-black bg-opacity-20 hover:bg-opacity-60"
                            type="button"
                          >
                            <span
                              class="anticon"
                            >
                              [iconify icon="mdi:pencil-outline"]
                            </span>
                          </button>
                        </div>
                        <div
                          class="ant-space-item"
                        >
                          <button
                            class="ant-btn ant-btn-circle ant-btn-text ant-btn-sm ant-btn-icon-only border-0 text-white text-opacity-80 hover:text-opacity-100 bg-black bg-opacity-20 hover:bg-opacity-60"
                            type="button"
                          >
                            <span
                              class="anticon"
                            >
                              [iconify icon="mdi:trash-can-outline"]
                            </span>
                          </button>
                        </div>
                      </div>
                    </div>
                  </div>
                </span>
              </span>
            </div>
            <div
              aria-grabbed="false"
              class="ant-tree-treenode ant-tree-treenode-draggable"
              draggable="true"
            >
              <span
                aria-hidden="true"
                class="ant-tree-indent"
              >
                <span
                  class="ant-tree-indent-unit ant-tree-indent-unit-start"
                />
              </span>
              <span
                class="ant-tree-switcher ant-tree-switcher-noop"
              />
              <span
                class="ant-tree-node-content-wrapper ant-tree-node-content-wrapper-normal"
                title="panel-01"
              >
                <span
                  class="ant-tree-title"
                >
                  <div
                    class="flex group"
                  >
                    <span>
                      panel-01
                    </span>
                    <div
                      class="opacity-0 group-hover:opacity-100 ml-2"
                    >
                      <div
                        class="ant-space ant-space-horizontal ant-space-align-center"
                      >
                        <div
                          class="ant-space-item"
                          style="margin-right: 8px;"
                        >
                          <button
                            class="ant-btn ant-btn-circle ant-btn-text ant-btn-sm ant-btn-icon-only border-0 text-white text-opacity-80 hover:text-opacity-100 bg-black bg-opacity-20 hover:bg-opacity-60"
                            type="button"
                          >
                            <span
                              class="anticon"
                            >
                              [iconify icon="mdi:pencil-outline"]
                            </span>
                          </button>
                        </div>
                        <div
                          class="ant-space-item"
                        >
                          <button
                            class="ant-btn ant-btn-circle ant-btn-text ant-btn-sm ant-btn-icon-only border-0 text-white text-opacity-80 hover:text-opacity-100 bg-black bg-opacity-20 hover:bg-opacity-60"
                            type="button"
                          >
                            <span
                              class="anticon"
                            >
                              [iconify icon="mdi:trash-can-outline"]
                            </span>
                          </button>
                        </div>
                      </div>
                    </div>
                  </div>
                </span>
              </span>
            </div>
            <div
              aria-grabbed="false"
              class="ant-tree-treenode ant-tree-treenode-leaf-last ant-tree-treenode-draggable"
              draggable="true"
            >
              <span
                aria-hidden="true"
                class="ant-tree-indent"
              >
                <span
                  class="ant-tree-indent-unit ant-tree-indent-unit-start"
                />
              </span>
              <span
                class="ant-tree-switcher ant-tree-switcher-noop"
              />
              <span
                class="ant-tree-node-content-wrapper ant-tree-node-content-wrapper-normal"
                title="panel-02"
              >
                <span
                  class="ant-tree-title"
                >
                  <div
                    class="flex group"
                  >
                    <span>
                      panel-02
                    </span>
                    <div
                      class="opacity-0 group-hover:opacity-100 ml-2"
                    >
                      <div
                        class="ant-space ant-space-horizontal ant-space-align-center"
                      >
                        <div
                          class="ant-space-item"
                          style="margin-right: 8px;"
                        >
                          <button
                            class="ant-btn ant-btn-circle ant-btn-text ant-btn-sm ant-btn-icon-only border-0 text-white text-opacity-80 hover:text-opacity-100 bg-black bg-opacity-20 hover:bg-opacity-60"
                            type="button"
                          >
                            <span
                              class="anticon"
                            >
                              [iconify icon="mdi:pencil-outline"]
                            </span>
                          </button>
                        </div>
                        <div
                          class="ant-space-item"
                        >
                          <button
                            class="ant-btn ant-btn-circle ant-btn-text ant-btn-sm ant-btn-icon-only border-0 text-white text-opacity-80 hover:text-opacity-100 bg-black bg-opacity-20 hover:bg-opacity-60"
                            type="button"
                          >
                            <span
                              class="anticon"
                            >
                              [iconify icon="mdi:trash-can-outline"]
                            </span>
                          </button>
                        </div>
                      </div>
                    </div>
                  </div>
                </span>
              </span>
            </div>
            <div
              aria-grabbed="false"
              class="ant-tree-treenode ant-tree-treenode-switcher-open ant-tree-treenode-leaf-last ant-tree-treenode-draggable"
              draggable="true"
            >
              <span
                aria-hidden="true"
                class="ant-tree-indent"
              />
              <span
                class="ant-tree-switcher ant-tree-switcher_open"
              >
                <span
                  aria-label="caret-down"
                  class="anticon anticon-caret-down ant-tree-switcher-icon"
                  role="img"
                >
                  <svg
                    aria-hidden="true"
                    data-icon="caret-down"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="0 0 1024 1024"
                    width="1em"
                  >
                    <path
                      d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                    />
                  </svg>
                </span>
              </span>
              <span
                class="ant-tree-node-content-wrapper ant-tree-node-content-wrapper-open"
                title="section-2"
              >
                <span
                  class="ant-tree-title"
                >
                  <div
                    class="flex group"
                  >
                    <span>
                      section-2
                    </span>
                    <div
                      class="opacity-0 group-hover:opacity-100 ml-2"
                    >
                      <div
                        class="ant-space ant-space-horizontal ant-space-align-center"
                      >
                        <div
                          class="ant-space-item"
                          style="margin-right: 8px;"
                        >
                          <button
                            class="ant-btn ant-btn-circle ant-btn-text ant-btn-sm ant-btn-icon-only border-0 text-white text-opacity-80 hover:text-opacity-100 bg-black bg-opacity-20 hover:bg-opacity-60"
                            type="button"
                          >
                            <span
                              class="anticon"
                            >
                              [iconify icon="mdi:pencil-outline"]
                            </span>
                          </button>
                        </div>
                        <div
                          class="ant-space-item"
                        >
                          <button
                            class="ant-btn ant-btn-circle ant-btn-text ant-btn-sm ant-btn-icon-only border-0 text-white text-opacity-80 hover:text-opacity-100 bg-black bg-opacity-20 hover:bg-opacity-60"
                            type="button"
                          >
                            <span
                              class="anticon"
                            >
                              [iconify icon="mdi:trash-can-outline"]
                            </span>
                          </button>
                        </div>
                      </div>
                    </div>
                  </div>
                </span>
              </span>
            </div>
            <div
              aria-grabbed="false"
              class="ant-tree-treenode ant-tree-treenode-draggable"
              draggable="true"
            >
              <span
                aria-hidden="true"
                class="ant-tree-indent"
              >
                <span
                  class="ant-tree-indent-unit ant-tree-indent-unit-end"
                />
              </span>
              <span
                class="ant-tree-switcher ant-tree-switcher-noop"
              />
              <span
                class="ant-tree-node-content-wrapper ant-tree-node-content-wrapper-normal"
                title="panel-11"
              >
                <span
                  class="ant-tree-title"
                >
                  <div
                    class="flex group"
                  >
                    <span>
                      panel-11
                    </span>
                    <div
                      class="opacity-0 group-hover:opacity-100 ml-2"
                    >
                      <div
                        class="ant-space ant-space-horizontal ant-space-align-center"
                      >
                        <div
                          class="ant-space-item"
                          style="margin-right: 8px;"
                        >
                          <button
                            class="ant-btn ant-btn-circle ant-btn-text ant-btn-sm ant-btn-icon-only border-0 text-white text-opacity-80 hover:text-opacity-100 bg-black bg-opacity-20 hover:bg-opacity-60"
                            type="button"
                          >
                            <span
                              class="anticon"
                            >
                              [iconify icon="mdi:pencil-outline"]
                            </span>
                          </button>
                        </div>
                        <div
                          class="ant-space-item"
                        >
                          <button
                            class="ant-btn ant-btn-circle ant-btn-text ant-btn-sm ant-btn-icon-only border-0 text-white text-opacity-80 hover:text-opacity-100 bg-black bg-opacity-20 hover:bg-opacity-60"
                            type="button"
                          >
                            <span
                              class="anticon"
                            >
                              [iconify icon="mdi:trash-can-outline"]
                            </span>
                          </button>
                        </div>
                      </div>
                    </div>
                  </div>
                </span>
              </span>
            </div>
            <div
              aria-grabbed="false"
              class="ant-tree-treenode ant-tree-treenode-leaf-last ant-tree-treenode-draggable"
              draggable="true"
            >
              <span
                aria-hidden="true"
                class="ant-tree-indent"
              >
                <span
                  class="ant-tree-indent-unit ant-tree-indent-unit-end"
                />
              </span>
              <span
                class="ant-tree-switcher ant-tree-switcher-noop"
              />
              <span
                class="ant-tree-node-content-wrapper ant-tree-node-content-wrapper-normal"
                title="panel-12"
              >
                <span
                  class="ant-tree-title"
                >
                  <div
                    class="flex group"
                  >
                    <span>
                      panel-12
                    </span>
                    <div
                      class="opacity-0 group-hover:opacity-100 ml-2"
                    >
                      <div
                        class="ant-space ant-space-horizontal ant-space-align-center"
                      >
                        <div
                          class="ant-space-item"
                          style="margin-right: 8px;"
                        >
                          <button
                            class="ant-btn ant-btn-circle ant-btn-text ant-btn-sm ant-btn-icon-only border-0 text-white text-opacity-80 hover:text-opacity-100 bg-black bg-opacity-20 hover:bg-opacity-60"
                            type="button"
                          >
                            <span
                              class="anticon"
                            >
                              [iconify icon="mdi:pencil-outline"]
                            </span>
                          </button>
                        </div>
                        <div
                          class="ant-space-item"
                        >
                          <button
                            class="ant-btn ant-btn-circle ant-btn-text ant-btn-sm ant-btn-icon-only border-0 text-white text-opacity-80 hover:text-opacity-100 bg-black bg-opacity-20 hover:bg-opacity-60"
                            type="button"
                          >
                            <span
                              class="anticon"
                            >
                              [iconify icon="mdi:trash-can-outline"]
                            </span>
                          </button>
                        </div>
                      </div>
                    </div>
                  </div>
                </span>
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
